third_party/move/move-compiler-v2/src/env_pipeline/model_ast_lints/blocks_in_conditions.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module implements an expression linter that checks for use of blocks
+//! in conditions. Such usage can make code harder to read, so we warn against it.
+//!
+//! Note that we do allow the use of blocks in conditions if there are inline
+//! specifications in them, as this is a common pattern to provide loop invariants.
+//!
+//! We also only report on the outermost condition with blocks.
+
+use crate::env_pipeline::model_ast_lints::ExpressionLinter;
+use codespan_reporting::diagnostic::Severity;
+use move_model::{
+    ast::ExpData,
+    model::{GlobalEnv, NodeId},
+};
+
+/// Expression linter keeping track of traversal state.
+#[derive(Default)]
+pub struct BlocksInConditions {
+    /// Is `None` if we are not traversing a condition with blocks.
+    state: Option<CondExprState>,
+}
+
+/// State of the traversal when a condition expression of interest has been found.
+enum CondExprState {
+    // Condition with `id` should be examined for the presence of spec blocks.
+    Examine { id: NodeId },
+    // Traversing within a condition with `id`.
+    // `has_spec_block` is true if we have seen a spec block within the condition so far.
+    Traversing { id: NodeId, has_spec_block: bool },
+}
+
+impl ExpressionLinter for BlocksInConditions {
+    fn visit_expr_pre(&mut self, expr: &ExpData, _env: &GlobalEnv) {
+        use ExpData::{Block, IfElse, Match, Sequence, SpecBlock};
+        match self.state {
+            None => {
+                if let IfElse(_, cond, _, _) | Match(_, cond, _) = expr {
+                    if let Block(id, ..) | Sequence(id, ..) = cond.as_ref() {
```

Comment:
This only looks at top-level expressions.  So it will miss something like:
```
    if { x = x + 1; x == 3 } && { y = y + 1; y == 2} {
    }
```
Or is that intentional?

Reply:
Nice catch. Note that the exact syntax you have provided is not valid Move (`if`'s condition needs to be wrapped in `()`), but the point you made was valid.

I have addressed this issue, and updated this lint, added corresponding tests.